### ISSUE TITLE
Add self-play training for Pong (replace rule-based opponent with policy snapshots)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#47]
## Task Type
Feature

## Task Description
Current Pong training (`examples/games/pong/train_pong.rs`) plays against a rule-based opponent that tracks the ball at 60% of the agent's speed (see `OPPONENT_SPEED` in `src/env/games/pong.rs:11`). The trained policy learns to rally indefinitely (episodes truncate at 2000 steps, avg 1986) but never has to beat a stronger opponent, so the skill ceiling is bounded by the hand-coded heuristic.

Add a **self-play training variant** where the right paddle is controlled by frozen policy snapshots periodically copied from the live agent. This is additive — `train_pong.rs` stays as the rule-based baseline.

## Out of Scope
- **Do NOT refactor `examples/games/pong/train_pong.rs`** — leave the rule-based variant as a baseline for comparison.
- Do NOT change the default web demo model loaded by `web/src/pages/PongPage.tsx` in this PR. The new model file just needs to exist; wiring the page to prefer it is a separate task.
- No changes to `src/multi_agent/matchmaking.rs` itself — `SelfPlayMatchmaker` already exists (lines 184–215) and works as-is.

## Acceptance Criteria
- [ ] New example file at `examples/games/pong/train_pong_self_play.rs`.
- [ ] New `[[example]]` entry in `Cargo.toml` (under the "Pong examples" section near line 152) named `train_pong_self_play`, with `required-features = ["training"]`.
- [ ] Pong env exposes a way to drive **both** paddles from external actions. Either:
  - **Preferred:** add `Pong::step_two(left_action: i64, right_action: i64) -> StepResult` that mirrors `step()` but skips the rule-based right-paddle update; OR
  - Add an `opponent_action` field/setter that the existing `step()` consults instead of the heuristic when set.
  - The existing `Environment::step(action)` (and therefore `train_pong.rs`) must keep its current behavior bit-for-bit — verified by the existing `train_pong` example still building and producing identical episode lengths on a smoke run.
- [ ] Observation mirroring helper for the right paddle: a free function or method that takes the 6D obs and returns the right-paddle perspective. The current obs is `[ball_x, ball_y, ball_dx, ball_dy, left_y, right_y]` (all normalized to [-1, 1]); the right-paddle view should negate `ball_x` and `ball_dx` and swap `left_y`/`right_y` so the same policy network can play either side.
- [ ] Training loop maintains a **snapshot pool** of recent frozen opponents (size ~5–10). Implementation guidance: a `Vec<MlpPolicy>` on CPU, FIFO eviction, push a clone every `SNAPSHOT_INTERVAL` updates (suggest 20–50). Initial pool seeded with the freshly-initialized random policy so episode 0 has a valid opponent.
- [ ] Each rollout episode samples one snapshot uniformly at random from the pool (per env, not per step). Use `rand` for sampling, matching the style in `SelfPlayMatchmaker` (`StdRng::from_entropy()`).
- [ ] **Logged metrics include opponent-pool win rate** over the last N episodes (e.g., wins ÷ total games where game = ball goes past one paddle). Print every 20 updates alongside the existing PPO stats. This is the primary signal that learning is happening.
- [ ] Final model exports as `pong_self_play_model.json` (and `.pt`) with `algorithm: "PPO + self-play"` in `TrainingMetadata`, hyperparameters including `snapshot_pool_size` and `snapshot_interval`, and notes mentioning self-play.
- [ ] Snapshot checkpoints saved to disk on each pool push (e.g., `pong_snapshot_{update}.pt`) so future evaluation tooling can replay matches between snapshots.

### Env API (minimal change)
Adding a second entry point to `Pong` is cleanest and avoids touching the trait. Sketch:

```rust
// in src/env/games/pong.rs, alongside the existing impl Pong block
pub fn step_two(&mut self, left_action: i64, right_action: i64) -> StepResult {
    // left paddle: same as in step()
    // right paddle: replace the OPPONENT_SPEED tracking block with the same
    //   match action { 0 => up, 2 => down, _ => stay } pattern
    // ball physics, collisions, scoring, termination: identical to step()
}

/// Return the obs from the right paddle's perspective: mirror x-axis, swap paddle positions.
pub fn mirror_observation(obs: &[f32]) -> Vec<f32> {
    vec![-obs[0], obs[1], -obs[2], obs[3], obs[5], obs[4]]
}
```

Resist the urge to extract the shared ball/collision logic into a private helper in the same PR — keeping `step` byte-identical to today is the easiest way to guarantee no regression. A follow-up can dedupe.

### Wiring into matchmaking infrastructure
`MatchmakingStrategy::SelfPlay` and `SelfPlayMatchmaker` already exist (`src/multi_agent/matchmaking.rs:55, 184-215`). **However**, the full `multi_agent` stack (`GameSimulator`, `PolicyLearner`, `Population`) is overkill for a 2-paddle Pong setup and `GameSimulator::obs_to_tensor` is still a placeholder (`src/multi_agent/simulator.rs:245`). For this issue, **do not use `GameSimulator`** — instead, follow the lighter pattern in `examples/games/snake/train_snake_multi_v2.rs`:

1. Single live `MlpPolicy` + optimizer (same shape as `train_pong.rs`).
2. A `Vec<MlpPolicy>` snapshot pool you manage in the training loop.
3. `EnvPool<Pong>` (or a `Vec<Pong>`) sized `NUM_ENVS`. For each env, at episode start, sample one snapshot from the pool and record `env_id -> snapshot_idx`.
4. In the per-step loop: build left-paddle obs batch + right-paddle (mirrored) obs batch, call the live policy on the left batch and the assigned snapshot on the right batch (you can group by snapshot_idx to keep the right-side batched), then call `step_two`.
5. Only the left-paddle transitions go into the rollout buffer (the right paddle is frozen — no gradient).
6. On episode end (scoring out or truncate), increment per-snapshot win/loss counters from the score, then re-sample a snapshot for that env.
7. Every `SNAPSHOT_INTERVAL` PPO updates: `let snap = live_policy.clone_for_inference()` (or save-then-load into a fresh `MlpPolicy`), push into pool, evict oldest if `pool.len() > POOL_MAX`.

### Seeding the initial pool
Before training starts, push **one clone of the freshly-initialized random policy** into the pool so update 0 has something to play against. Do NOT pretrain against the rule-based opponent — the whole point is to measure self-play learning curve from scratch.

### Suggested hyperparameters (starting point)
- `NUM_ENVS = 32`, `NUM_STEPS = 128` (match `train_pong.rs`)
- `TOTAL_TIMESTEPS = 20_000_000` (same)
- `SNAPSHOT_INTERVAL = 25` updates (~3.2M steps between pool pushes)
- `POOL_MAX = 8` snapshots
- Other PPO config identical to `train_pong.rs`

### Cloning `MlpPolicy`
`MlpPolicy::save`/`load` exist (`src/policy/mlp.rs:235, 241`). The simplest snapshot path: `policy.save("tmp_snap.pt")` → `let mut snap = MlpPolicy::new(...); snap.load("tmp_snap.pt")`. If a direct in-memory clone is awkward via `tch::nn::VarStore`, save/load through a temp file is acceptable for the snapshot cadence (every ~25 updates, not hot).

## Files to Touch
- **New:** `examples/games/pong/train_pong_self_play.rs` (~300 lines, modeled on `train_pong.rs` + the snapshot bookkeeping)
- **Modify:** `src/env/games/pong.rs` — add `step_two` and `mirror_observation` (additive, no changes to existing `step`)
- **Modify:** `Cargo.toml` — add `[[example]]` entry near line 156
- **Untouched:** `train_pong.rs`, `src/multi_agent/*`, `web/`

## Test / Smoke Plan
1. `cargo build --example train_pong_self_play --release --features training` — compiles.
2. `cargo build --example train_pong --release --features training` — existing example still compiles unchanged.
3. **Unit test in `src/env/games/pong.rs`:** Construct a `Pong`, call `step_two(1, 1)` (both stay) for many steps and verify ball physics match `step(1)` for the first N steps where the rule-based right paddle would also be at center (or just verify that with `step_two`, the right paddle stays put when the action is "stay" while the ball moves — confirming the rule-based heuristic is bypassed).
4. **Unit test for `mirror_observation`:** Round-trip — `mirror(mirror(obs)) == obs`.
5. **Smoke training run:** Run `train_pong_self_play` for ~500K steps (override `TOTAL_TIMESTEPS` with an env var or short-run constant) and confirm:
   - It compiles, runs, and exits cleanly.
   - The opponent-pool win rate log line appears and is non-trivial (not stuck at 0% or 100%).
   - `pong_self_play_model.json` is written with the correct `algorithm` field.
6. Manual sanity: after a short run, load `pong_self_play_model.json` in the web demo (drop it in place of `pong_model.json`) and visually confirm the agent doesn't immediately collapse — it should at least track the ball.

## Technical Notes
- The current `Pong::step` hard-codes the rule-based right paddle via `OPPONENT_SPEED` (`src/env/games/pong.rs:11, 108-109`). The two-paddle path must not regress single-player training — keep both modes available.
- Observation is the same 6D vector regardless of side; the mirror helper handles perspective swap so one policy network plays both paddles.
- Snapshot eviction is FIFO — newest snapshot is the strongest opponent (assuming monotonic learning), so a uniform sample over the pool gives a curriculum from "weakest historical" to "current minus one." This is the standard PPO + self-play recipe (PFSP/league play is a future extension, not needed here).
- The existing `MULTI_AGENT_DESIGN.md` documents the full population/league architecture (`Population`, `GameSimulator`, `PolicyLearner`) but this issue intentionally stays simpler — a single live learner with a frozen snapshot pool is the right granularity for 2-player Pong and matches the snake `train_snake_multi_v2.rs` pattern.

**Comment by rjwalters** [comment]
**Curator notes — framing summary**

Surveyed the repo and clarified scope so this issue is implementation-ready:

- **Additive only**: `train_pong.rs` and the existing `Pong::step` rule-based opponent stay untouched as a baseline. The new training script lives at `examples/games/pong/train_pong_self_play.rs`.
- **Lighter than the full multi-agent stack**: `MatchmakingStrategy::SelfPlay` exists at `src/multi_agent/matchmaking.rs:55`, but the full `GameSimulator` / `PolicyLearner` machinery is overkill for 2-paddle Pong (and `GameSimulator::obs_to_tensor` is still a placeholder). The Builder should model the loop on `examples/games/snake/train_snake_multi_v2.rs` instead — single live policy + a `Vec<MlpPolicy>` snapshot pool, FIFO eviction, uniform sampling, ~25-update cadence.
- **Env change is minimal and additive**: add `Pong::step_two(left, right)` and a `mirror_observation` helper. Existing `Environment::step(action)` must remain bit-identical.
- **Win-rate vs the snapshot pool is the load-bearing metric** — without it, you can't tell if self-play is actually improving the policy.
- Pool is seeded with one clone of the random init so update 0 has an opponent; no pretraining against the rule-based heuristic.

Out-of-scope and called out explicitly: no refactor of `train_pong.rs`, no changes to the web demo's default model loading, no changes to `matchmaking.rs` itself.

Acceptance criteria, files-to-touch, smoke plan, and hyperparameter defaults are now in the issue body.
